### PR TITLE
fix: collapse sidenav menus when side nav is closed

### LINKED ISSUE #15900: [Bug]: SideNavMenu remains expanded, when Rail SideNav collapses
### Package

@carbon/react

### Browser

Edge

### Package version

1.51.0

### React version

18.2.0

### Description

SideNavMenu remains expanded, when Rail SideNav collapses

image

### Reproduction/example

https://react.carbondesignsystem.com/?path=/story/components-ui-shell-sidenav--side-nav-rail-w-header

### Steps to reproduce

1. Hover cursor over SideNav for it to expand
2. In expanded SideNav expand first SideNavMenu
3. Move cursor away from SideNav for it to collapse into rail
4. Now you should see that there is additional empty space between first item and second item in SideNav

### Suggested Severity

Severity 3 = User can complete task, and/or has a workaround within the user experience of a given component.

### Application/PAL

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [X] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

### PR DESCRIPTION
**Pull request by riddhybansal**
Closes #15900 

Submenus should close when side nav closes

#### Changelog

**Changed**

- Style changes

#### Testing / Reviewing

Go to storybook , Open Side Nav -> Open Submenu -> Close side nav ,  sub menu  should collapse autromatically.

### CODE REVIEW

**Review by tay1orjones — CHANGES_REQUESTED** [review]
Generally `display: none` can't be used because it [removes the element(s) from the accessibility tree](https://developer.mozilla.org/en-US/docs/Web/CSS/display#accessibility_concerns) and so it won't be read by screenreaders.  

Instead of a style-related fix, would it make sense to modify the internal state of SideNavMenu to account for the change in the parent SideNav?

This bug was introduced https://github.com/carbon-design-system/carbon/pull/15349, my guess is the fix for this would be something similar to what was removed, but without reintroducing the bug it was fixing.

File: packages/react/src/components/UIShell/SideNavMenu.tsx [hunk 1]
```diff
@@ -94,6 +95,18 @@ const SideNavMenu = React.forwardRef<HTMLElement, SideNavMenuProps>(
       [customClassName as string]: !!customClassName,
     });
 
+    if (isSideNavExpanded === false && isExpanded === true && isRail === true) {
```
Comment (tay1orjones) [inline packages/react/src/components/UIShell/SideNavMenu.tsx]:
```suggestion
    if (!isSideNavExpanded && isExpanded && isRail) {
```

File: packages/react/src/components/UIShell/SideNavMenu.tsx [hunk 2]
```diff
@@ -94,6 +95,18 @@ const SideNavMenu = React.forwardRef<HTMLElement, SideNavMenuProps>(
...
+      setIsExpanded(false);
+      setPrevExpanded(true);
+    } else if (
+      isSideNavExpanded === true &&
+      prevExpanded === true &&
+      isRail === true
+    ) {
```
Comment (tay1orjones) [inline packages/react/src/components/UIShell/SideNavMenu.tsx]:
```suggestion
    } else if (isSideNavExpanded && prevExpanded && isRail) {
```

**Review by tay1orjones — CHANGES_REQUESTED** [review]

**Review by tay1orjones — APPROVED** [review]

**Review by guidari — APPROVED** [review]
LGTM! 🚀

**Pull request merged**